kobo-uncaged/kunc/kunc.go

```diff
@@ -0,0 +1,296 @@
+		}
+	} else {
+		for _, kmd := range ku.k.MetadataMap {
+			md := map[string]interface{}{}
+			//spew.Dump(kmd)
+			mapstructure.Decode(kmd, &md)
+			mdList = append(mdList, md)
+		}
+	}
+	return mdList
+}
+
+// GetDeviceInfo asks the client for information about the drive info to use
+func (ku *koboUncaged) GetDeviceInfo() uc.DeviceInfo {
+	return ku.k.DriveInfo
+}
+
+// SetDeviceInfo sets the new device info, as comes from calibre. Only the nested
+// struct DevInfo is modified.
+func (ku *koboUncaged) SetDeviceInfo(devInfo uc.DeviceInfo) {
+	ku.k.DriveInfo = devInfo
+	ku.k.SaveDeviceInfo()
+}
+
+// UpdateMetadata instructs the client to update their metadata according to the
+// new slice of metadata maps
+func (ku *koboUncaged) UpdateMetadata(mdList []map[string]interface{}) {
+	for _, md := range mdList {
+		koboMD := device.CreateKoboMetadata()
+		mapstructure.Decode(md, &koboMD)
+		koboMD.Thumbnail = nil
+		cid := util.LpathToContentID(koboMD.Lpath, string(ku.k.ContentIDprefix))
+		ku.k.MetadataMap[cid] = koboMD
+		ku.k.UpdatedMetadata = append(ku.k.UpdatedMetadata, cid)
+	}
+	ku.k.WriteMDfile()
+	ku.k.WriteUpdateMDfile()
+}
+
+// GetPassword gets a password from the user.
+func (ku *koboUncaged) GetPassword(calibreInfo uc.CalibreInitInfo) string {
+	return ku.k.Passwords.NextPassword()
+}
+
+// GetFreeSpace reports the amount of free storage space to Calibre
+func (ku *koboUncaged) GetFreeSpace() uint64 {
+	// Note, this method of getting available disk space is Linux specific...
+	// Don't try to run this code on Windows. It will probably fall over
+	var fs syscall.Statfs_t
+	err := syscall.Statfs(ku.k.BKRootDir, &fs)
+	if err != nil {
+		log.Println(err)
+		// Despite the error, we return an arbitrary amount. Thoughts on this?
+		return 1024 * 1024 * 1024
+	}
+	return fs.Bavail * uint64(fs.Bsize)
+}
+
+// SaveBook saves a book with the provided metadata to the disk.
+// Implementations return an io.WriteCloser (book) for UNCaGED to write the ebook to
+// lastBook informs the client that this is the last book for this transfer
+// newLpath informs UNCaGED of an Lpath change. Use this if the lpath field in md is
+// not valid (eg filesystem limitations.). Return an empty string if original lpath is valid
+func (ku *koboUncaged) SaveBook(md map[string]interface{}, len int, lastBook bool) (book io.WriteCloser, newLpath string, err error) {
+	koboMD := device.CreateKoboMetadata()
+	mapstructure.Decode(md, &koboMD)
+	// The calibre wireless driver does not sanitize the filepath for us. We sanitize it here,
+	// and if lpath changes, inform Calibre of the new lpath.
+	newLpath = ku.k.InvalidCharsRegex.ReplaceAllString(koboMD.Lpath, "_")
+	// Also, for kepub files, Calibre defaults to using "book/path.kepub"
+	// but we require "book/path.kepub.epub". We change that here if needed.
+	newLpath = util.LpathKepubConvert(newLpath)
+	if newLpath != koboMD.Lpath {
+		koboMD.Lpath = newLpath
+	} else {
+		newLpath = ""
+	}
```

Comment:
I would replace this with:

```go
if nlp := util.LpathKepubConvert(newLpath); nlp != koboMD.Lpath {
    newLpath = nlp
} else {
    nlp = ""
}
```

Reply:
Almost there. `newLpath` is returned, so it needs to be an empty string if `nlp == koboMD.Lpath`

Reply:
Sorry, meant for the second variable to be newLpath.

Reply:
Yeah, I figured that's what you meant, and what I implemented.